### ISSUE TITLE
Use reva token for auth

### ISSUE THREAD

**Issue by tbsbdr** [owncloud/web#9245]
# Description

## User Stories

* > As an admin I want to control the session lifetime in the client so that I'm not dependent on the IAM.

## Definition of ready
[ ] everybody needs to understand the value written in the user story
[ ] acceptance criteria has to be defined
[ ] all dependencies of the user story need to be identified
[ ] feature should be seen from an end user perspective
[ ] user story has to be estimated
[ ] story points need to be less then 20

## Definition of done
- Functional requirements
[ ] functionality described in the user story works
[ ] acceptance criteria are fulfilled
- Quality
[ ] code review happened
[ ] CI is green
[ ] critical code received unit tests by the developer
[ ] automated tests passed (if automated tests are not available, this test needs to be created and passed
- Non-functional requirements
[ ] no sonar cloud issues

realtes to https://github.com/owncloud/web/pull/9159/commits/9c37d8cfa80a7fd3dafc50e14df8647ea1cf3680

**Comment by dragotin** [comment]
Proposal how to implement: 

The extension system allows to register custom code in a post-sign-in-hook. There, the CERN code could be added to an extension that pulls the reva token and stores it in the browser.

Benefit for the product would for example be that an extension could be written that lists all logged in users on a dashboard. For that, the custom extension code would call an dashboard endpoint with the user  name that just signed in.

**Comment by tbsbdr** [comment]
see same example at mattermost

**Comment by tbsbdr** [comment]
@DeepDiver1975 this patch https://github.com/owncloud/web/commit/9c37d8cfa80a7fd3dafc50e14df8647ea1cf3680 from CERN uses the REVA token so that the session lifetime can be controlled via the client. Can you assess if this option would be a security issue for the product?  (Generally we wonder if this would be an option for the product.). thank you!

**Comment by tbsbdr** [comment]
Thx! Fyi @hodyroff @labkode

**Comment by kulmann** [comment]
Let me rephrase @tbsbdr question:

The web patch uses a kind of post login hook to use the just freshly created access token from the IdP to make an authenticated request to a reva endpoint which then generates a custom token (session life time is still controlled by the backend). This custom token is then used instead of the original access token  in all following requests from ownCloud Web. The original access token is being sent into nirvana.

Please check if you need to reevaluate based on this @DeepDiver1975 :) 

Regardless of the outcome, the possibility of a post login hook via extension system is a good idea.

**Comment by DeepDiver1975** [comment]
That is an interesting detail which I need to think about. 
Does this mean that the refresh token is also thrown away? 

The functional impact needs to be thought about.

Overall my comment from about remains.....

**Comment by kulmann** [comment]
> Does this mean that the refresh token is also thrown away?

The particular setup doesn't allow to request the `offline_access` scope for web from what I know. So there is no refresh token that could be thrown away.

Some context again: The particular SSO team doesn't allow a session lifetime (neither through a long IdP session timeout nor via a refresh token) that is long enough to make their users happy, so engineering found an alternative solution.

**Comment by DeepDiver1975** [comment]
Does that mean tokens have no expiry in this environment? Anyway ....

**Comment by kulmann** [comment]
> Does that mean tokens have no expiry in this environment? Anyway ....

No idea how the tokens behave client side. I just know the flow how to request such a custom token. You can ping @tbsbdr if you want to take a look at the requests in the browser, he has a test user.

**Comment by labkode** [comment]
Hi, 

Any decent web application will have session control independently of authentication source (OIDC, basic auth, LDAP, ...).

Here you can find two examples:

https://github.com/mattermost/mattermost/blob/c6916c7047507eb6915e2ed047d3e3688cf3694a/server/channels/app/oauth.go#L396

```
    session.AddProp(model.SessionPropOs, "OAuth2")
    session.AddProp(model.SessionPropBrowser, "OAuth2")

    session, err := a.Srv().Store().Session().Save(session)
    if err != nil {
        return nil, model.NewAppError("newSession", "api.oauth.get_access_token.internal_session.app_error", nil, "", http.StatusInternalServerError)
    }
```

[indico/indico/indico/modules/auth/init.py](https://github.com/indico/indico/blob/ad619cd5546919703702e8a4795f5e3102faa39e/indico/modules/auth/__init__.py#L74)
```
        identity.data = identity_info.data
    if user.is_blocked:
        raise MultipassException(_('Your Indico profile has been blocked.'))
    login_user(user, identity)


def login_user(user, identity=None, admin_impersonation=False):
```

**Comment by JammingBen** [comment]
Implemented via https://github.com/owncloud/web/pull/9871.
